runtime/interpreter/interpreter.go

```diff
@@ -2573,11 +2573,188 @@ func init() {
+			func(invocation Invocation) Value {
+				typeID := invocation.Arguments[0].(*StringValue).Str
+				location, qualifiedIdentifier, err := common.DecodeTypeID(typeID)
+
+				// if the typeID is invalid, return nil
+				if err != nil || location == nil {
+					return NilValue{}
+				}
+
+				interfaceType, err := invocation.Interpreter.getInterfaceType(location, qualifiedIdentifier)
+				if err != nil {
+					return NilValue{}
+				}
+
+				return NewSomeValueOwningNonCopying(TypeValue{
+					Type: ConvertSemaToStaticType(interfaceType),
+				})
+			},
+			sema.InterfaceTypeFunctionType,
+		),
+	)
+
+	defineBaseValue(
+		baseActivation,
+		"FunctionType",
+		NewHostFunctionValue(
+			func(invocation Invocation) Value {
+				parameters := invocation.Arguments[0].(*ArrayValue).values
+				returnType := invocation.Interpreter.ConvertStaticToSemaType(invocation.Arguments[1].(TypeValue).Type)
+				parameterTypes := make([]*sema.Parameter, 0, len(parameters))
+				for _, param := range parameters {
+					parameterTypes = append(
+						parameterTypes,
+						&sema.Parameter{
+							TypeAnnotation: sema.NewTypeAnnotation(invocation.Interpreter.ConvertStaticToSemaType(param.(TypeValue).Type)),
+						},
+					)
+				}
+
+				return TypeValue{
+					Type: FunctionStaticType{
+						Type: &sema.FunctionType{
+							ReturnTypeAnnotation: sema.NewTypeAnnotation(returnType),
+							Parameters:           parameterTypes,
+						},
+					}}
+			},
+			sema.FunctionTypeFunctionType,
+		),
+	)
+
+	defineBaseValue(
+		baseActivation,
+		"RestrictedType",
+		NewHostFunctionValue(
+			RestrictedTypeFunction,
+			sema.RestrictedTypeFunctionType,
+		),
+	)
+}
+
+func RestrictedTypeFunction(invocation Invocation) Value {
+	restrictedIDs := invocation.Arguments[1].(*ArrayValue).values
+
+	staticRestrictions := make([]InterfaceStaticType, 0, len(restrictedIDs))
+	semaRestrictions := make([]*sema.InterfaceType, 0, len(restrictedIDs))
+	for _, typeID := range restrictedIDs {
+		location, qualifiedIdentifier, err := common.DecodeTypeID(typeID.(*StringValue).Str)
+		// if the typeID is invalid, return nil
+		if err != nil || location == nil {
+			return NilValue{}
+		}
+
+		restrictionInterface, err := invocation.Interpreter.getInterfaceType(location, qualifiedIdentifier)
+		if err != nil {
+			return NilValue{}
+		}
+
+		staticRestrictions = append(staticRestrictions, ConvertSemaToStaticType(restrictionInterface).(InterfaceStaticType))
+		semaRestrictions = append(semaRestrictions, restrictionInterface)
+	}
+
+	var semaType sema.Type
+
+	switch typeID := invocation.Arguments[0].(type) {
+	case NilValue:
+		semaType = nil
+	case *SomeValue:
+		location, qualifiedIdentifier, err := common.DecodeTypeID(typeID.Value.(*StringValue).Str)
+		// if the typeID is invalid, return nil
+		if err != nil || location == nil {
+			return NilValue{}
+		}
+		semaType, err = invocation.Interpreter.getCompositeType(location, qualifiedIdentifier)
+		if err != nil {
+			return NilValue{}
+		}
```

Comment:
Maybe refactor this logic here into a function, it's a duplicate of the code of the `CompositeType` constructor function. Like the constructor, it incorrectly rejects built-in types (`location == nil`)